File: betterrandom/src/main/java/io/github/pr0methean/betterrandom/util/CloneViaSerialization.java
package io.github.pr0methean.betterrandom.util;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;

public enum CloneViaSerialization {
  /* Utility class with no instances */;

  /**
   * Clones an object by serializing and deserializing it.
   * @param object The object to clone.
   * @param <T> The type of {@code object}.
   * @return A clone of {@code object}.
   */
  @SuppressWarnings("unchecked") public static <T extends Serializable> T clone(final T object) {
    byte[] serialCopy;
    serialCopy = toByteArray(object);
    // Read the object back-in.
    return fromByteArray(serialCopy);
  }

  public static <T extends Serializable> T fromByteArray(byte[] serialCopy) {
    try (ObjectInputStream objectInStream = new ObjectInputStream(
        new ByteArrayInputStream(serialCopy))) {
      return (T) (objectInStream.readObject());
    } catch (IOException | ClassNotFoundException e) {
      throw new RuntimeException(e);
    }
  }

  public static <T extends Serializable> byte[] toByteArray(T object) {
    byte[] serialCopy;
    try (ByteArrayOutputStream byteOutStream = new ByteArrayOutputStream();
        ObjectOutputStream objectOutStream = new ObjectOutputStream(byteOutStream)) {
      objectOutStream.writeObject(object);
      serialCopy = byteOutStream.toByteArray();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return serialCopy;
  }
}

File: betterrandom/src/test/java/io/github/pr0methean/CloneViaSerialization.kt
package io.github.pr0methean.betterrandom

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.io.Serializable

enum class CloneViaSerialization {
    ;

    companion object {

        /**
         * Clones an object by serializing and deserializing it.
         * @param object The object to clone.
         * @param <T> The type of `object`.
         * @return A clone of `object`.
        </T> */
        fun <T : Serializable> clone(`object`: T): T {
            return fromByteArray(toByteArray(`object`))
        }

        fun <T : Serializable> fromByteArray(serialCopy: ByteArray): T {
            try {
                ObjectInputStream(
                        ByteArrayInputStream(serialCopy)).use { objectInStream -> return objectInStream.readObject() as T }
            } catch (e: IOException) {
                throw RuntimeException(e)
            } catch (e: ClassNotFoundException) {
                throw RuntimeException(e)
            }

        }

        fun <T : Serializable> toByteArray(`object`: T): ByteArray {
            try {
                ByteArrayOutputStream().use { byteOutStream ->
                    ObjectOutputStream(byteOutStream).use { objectOutStream ->
                        objectOutStream.writeObject(`object`)
                        return byteOutStream.toByteArray()
                    }
                }
            } catch (e: IOException) {
                throw RuntimeException(e)
            }

        }
    }
}/* Utility class with no instances */
